src/modules/users/SetPasswordJob.cpp

```diff
@@ -44,6 +44,7 @@ SetPasswordJob::prettyStatusMessage() const
     return tr( "Setting password for user %1." ).arg( m_userName );
 }
 
+#ifndef HAVE_CRYPT_GENSALT
```

Comment:
I would ifdef the implementation, myself: leave the API intact, leave the callers unchanged, but
- fix the misleading comment (it uses `length`, not 16)
- change the comment into a full Doxygen `/** */` comment explaining that ... but if this other thing ...
- internally, use `#ifdef HAVE_CRYPT_GENSALT` to return `crypt_gensalt( NULL, 0, NULL, 0 )` 

Is there no way to specify a length for that function? Does it just always return 16 bytes, or what? 

Also, prefer `nullptr` over `NULL`.

Reply:
> Is there no way to specify a length for that function? Does it just always return 16 bytes, or what?

So I think this question comes from a misunderstanding of what the `crypt()` function actually is doing. It takes two arguments, `prefix`, and `password`, and returns a hash which is guaranteed to be identical to output of another call of the function with the same prefix and password. 

The prefix is composed of multiple parts. Previously, we were manually construction a prefix that would look like this: `$6$GkZUpcVNX.ErIfwo`. This breaks down to `$6` as specifying the `sha512crypt` algorithm, and then `GkZUpcVNX.ErIfwo` as the randomly generated salt to use. The `sha512crypt` algorithm only accepts two inputs, the salt and the password so this prefix was sufficient for the `crypt()` function.

What `crypt_gensalt( NULL, 0, NULL, 0 )` is doing instead, is asking libxcrypt to return a cryptographically generated prefix of whatever algorithm libxcrypt considers strongest of the algorithms it was compiled with and the default settings for that algorithm. libxcrypt is almost always compiled with the `--enable-hashes=strong` option, which means that it will consider yescrypt to be the strongest algorithm it has available. 

On such a system an example prefix returned by `crypt_gensalt` would be `$y$j9T$ixuQH9rz2Bfh9yWXs0BJ2/`. This breaks down to `$y` as the identifier for the yescrypt algorithm, `j9T` as parameters of the yescrypt algorithm (the breakdown of what `j9T` mean is a bit outside the scope of this response, but you can read up on it [here](https://unix.stackexchange.com/questions/690679/what-does-j9t-mean-in-yescrypt-from-etc-shadow) if you want). Then `ixuQH9rz2Bfh9yWXs0BJ2/` is a randomly generated salt suitable for cryptographic use. 

It is possible to change the arguments of the `crypt_gensalt` algorithm to get a specific algorithm or salt of a specified length out of it, but it's not really necessary (or even desirable) in this case. The entire point of switching to `crypt_gensalt` is to pass the decision of what algorithm and other default settings to the distro so they can use whatever they feel best (via patches or compile-time build parameters). Solus itself is specifically seeking this feature because we've migrated to yescrypt as the default password hashing algorithm and would like that to be reflected by hashes created by the installer as well.

Some background information on this from the related Fedora change: https://fedoraproject.org/wiki/Changes/yescrypt_as_default_hashing_method_for_shadow

Reply:
Thanks for the explanation.